README.md
Franges
===========

Franges adds support for floating point and fixed precision (Decimal) range generator functions.

Usage
=====

```python
#!/usr/bin/env python

from franges import drange
from franges import frange

for x in drange(0,2,0.1, precision = 6):
print(x)
for x in frange(0,2,0.1):
print(x)

list(drange(0,1,0.1, 6)) # [0, 0.1, 0.2, ... , 0.9]
list(frange(1,0,-0.2)) # [1.0, 0.8, 0.6, 0.3999999999999999, 0.19999999999999996]
list(drange(1,0,-0.2,6)) # [1.0, 0.8, 0.6, 0.4, 0.2]
```
    
Contributors
============
Nisan Haramati       hanisan@gmail.com


CHANGES.txt
v1.0.0, 2015-11-15 -- migrate to github and add tests.
v0.1.0, 2012-12-13 -- Initial Release.


tests/test_drange.py
#!/usr/bin/env python

from franges import drange


def test_basic_usage():
    expected = [0, 0.1, 0.2, 0.3]
    assert(expected == list(drange(0, 0.4, 0.1, 6)))
    expected = [0, 0.5, 1.0, 1.5]
    assert(expected == list(drange(0, 2, 0.5, 2)))


def test_partial_args():
    expected = list(range(10))
    assert(expected == list(drange(10)))


tests/test_frange.py
#!/usr/bin/env python

from franges import frange


def test_basic_usage():
    expected = [0, 0.5]
    assert(expected == list(frange(0, 1, 0.5)))
    expected = [0, 0.5, 1]
    assert(expected == list(frange(0, 1.5, 0.5)))


def test_partial_args():
    expected = list(range(10))
    assert(expected == list(frange(10)))


franges/drange.py
#
# Copyright (C) 2012  POF.com
#
# This file is part of franges.
#
# franges is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# franges is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with franges.  If not, see <http://www.gnu.org/licenses/>.

# drange.py
import decimal
from math import ceil
# create the best xrange function you can based on python
# version (2.7.x / 3.x.x)
try:
    _xrange = xrange
except NameError:
    _xrange = range


def drange(start, stop=None, step=1, precision=None):
    """drange generates a set of Decimal values over the
    range [start, stop) with step size step

    drange([start,] stop, [step [,precision]])"""

    if stop is None:
        for x in _xrange(int(ceil(start))):
            yield x
    else:
        # find precision
        if precision is not None:
            decimal.getcontext().prec = precision
        # convert values to decimals
        start = decimal.Decimal(start)
        stop = decimal.Decimal(stop)
        step = decimal.Decimal(step)
        # create a generator expression for the index values
        indices = (
            i for i in _xrange(
                0,
                int(((stop-start)/step).to_integral_value())
            )
        )

        # yield results
        for i in indices:
            yield float(start + step*i)


franges/frange.py
#
# Copyright (C) 2012  POF.com
#
# This file is part of franges.
#
# franges is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# franges is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with franges.  If not, see <http://www.gnu.org/licenses/>.

# frange.py
from math import ceil
# create the best xrange function you can based on python
# version (2.7.x / 3.x.x)
try:
    _xrange = xrange
except NameError:
    _xrange = range


def frange(start, stop=None, step=1):
    """frange generates a set of floating point values over the
    range [start, stop) with step size step

    frange([start,] stop [, step ])"""

    if stop is None:
        for x in _xrange(int(ceil(start))):
            yield x
    else:
        # create a generator expression for the index values
        indices = (i for i in _xrange(0, int((stop-start)/step)))
        # yield results
        for i in indices:
            yield start + step*i


franges/__init__.py
# 
# Copyright (C) 2012  POF.com
#
# This file is part of franges.
#
# franges is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# franges is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with franges.  If not, see <http://www.gnu.org/licenses/>.

"""Franges adds support for floating point and fixed precision (Decimal) range generator functions

Syntax is identical to range([start,] stop [,step]) for frange,
and similar for drange, with an additional precision parameter
frange([start,] stop [, step])
drange([start,] stop [,step [,precision]])

drange returns Decimal types and can enforce precision if specified. Otherwise it is identical to frange.

Sample usage:

from franges import drange
for x in drange(0, 2, 0.1, 6):
    ...
    
from franges import frange
for x in frange(0, 2, 0.1):
    ...

list(drange(0,1,0.1, 6)) # [0, 0.1, 0.2, ... , 0.9]
list(frange(1,0,-0.2)) # [1.0, 0.8, 0.6, 0.3999999999999999, 0.19999999999999996]
list(drange(1,0,-0.2,6)) # [1.0, 0.8, 0.6, 0.4, 0.2]
"""




from .drange import drange
from .frange import frange

__all__ = ['frange', 'drange']
